### PR TITLE
Shareable links backend - Models, APIs, bypass auth for table requests

### LINKED ISSUE #3033: Implement Django models and APIs for Shares
I've not added a description, since I'll be implementing this.

Refer [Product spec](https://wiki.mathesar.org/en/product/specs/publicly-shareable-links) for detailed info.

### PR DESCRIPTION
**Pull request by pavish**
Fixes #3033 
Related to #3034 

This PR:
* Adds Models for SharedTable and SharedQuery
* Adds APIs for SharedTable and SharedQuery
* Simplifies access control checks in all of tables' resources: columns, constraints, records
* Bypasses authentication for table requests for valid share uuids
* Adds custom common_data for shared tables required by frontend

Note:
* This PR only makes authentication related changes to tables. Queries will be handled in subsequent PRs.

Flows built in this PR:
* A table or query can be shared by posting to their respective endpoints:
  - `/api/ui/v0/tables/<table_pk>/shares`
  - `/api/ui/v0/queries/<query_pk>/shares`
* Update requests are allowed for shares. However only the `enabled` property is allowed to be updated.
* Requests with valid token to the following endpoints should succeed with a valid token:
    - `/api/db/v0/tables/<table_pk>/?shared-link-uuid=<uuid>`
    - `/api/db/v0/tables/<table_pk>/columns/?shared-link-uuid=<uuid>`
    - `/api/db/v0/tables/<table_pk>/constraints/?shared-link-uuid=<uuid>`
    - `/api/db/v0/tables/<table_pk>/records/?shared-link-uuid=<uuid>`
* The following user flows should work for the above endpoints:
    - Unauthenticated requests with valid `shared-link-uuid`
    - Authenticated requests made by users who do not have explicit access the table, but have valid `shared-link-uuid`
    - Authenticated requests made by users who have access to the table, with or without valid `shared-link-uuid`

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by pavish** [comment]
@dmos62 @silentninja 

The requests work well when I make them from the frontend.

However, there seems to be random errors when I make them from the Django browsable api view. I'm not sure why, but I'm concerned I've missed something or perhaps there's a bug.

For eg.,

This request succeeds `http://localhost:8000/api/db/v0/tables/1/constraints/?shared-link-uuid=874cfd2b-d580-4ea1-9ded-c226d46f9443` when made via the frontend.

But when made via the Django browseable API view, the following error is thrown:
```
Environment:


Request Method: GET
Request URL: http://localhost:8000/api/db/v0/tables/1/constraints/?shared-link-uuid=874cfd2b-d580-4ea1-9ded-c226d46f9443

Django Version: 3.1.14
Python Version: 3.9.17
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'whitenoise.runserver_nostatic',
 'django.contrib.staticfiles',
 'rest_framework',
 'django_filters',
 'django_property_filter',
 'drf_spectacular',
 'mathesar']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'whitenoise.middleware.WhiteNoiseMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'mathesar.middleware.CursorClosedHandlerMiddleware',
 'mathesar.middleware.PasswordChangeNeededMiddleware',
 'django_userforeignkey.middleware.UserForeignKeyMiddleware',
 'django_request_cache.middleware.RequestCacheMiddleware']



Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 204, in _get_response
    response = response.render()
  File "/usr/local/lib/python3.9/site-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File "/usr/local/lib/python3.9/site-packages/rest_framework/response.py", line 70, in rendered_content
    ret = renderer.render(self.data, accepted_media_type, context)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 724, in render
    context = self.get_context(data, accepted_media_type, renderer_context)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 702, in get_context
    'filter_form': self.get_filter_form(data, view, request),
  File "/usr/local/lib/python3.9/site-packages/rest_framework/renderers.py", line 634, in get_filter_form
    html = backend().to_html(request, queryset, view)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 304, in to_html
    context = self.get_template_context(request, queryset, view)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 296, in get_template_context
    for key, label in self.get_valid_fields(queryset, view, context):
  File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 250, in get_valid_fields
    return self.get_default_valid_fields(queryset, view, context)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 230, in get_default_valid_fields
    model_property_names = [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/filters.py", line 232, in <listcomp>
    attr for attr in dir(model_class) if isinstance(getattr(model_class, attr), property) and attr != 'pk'
  File "/code/mathesar/state/cached_property.py", line 56, in __get__
    key = self._get_ip_key(instance=instance)
  File "/code/mathesar/state/cached_property.py", line 82, in _get_ip_key
    ip_key = self._get_random_ip_key(instance)
  File "/code/mathesar/state/cached_property.py", line 111, in _get_random_ip_key
    ip_key = self._get_ip_key_from_instance_cache(instance)
  File "/code/mathesar/state/cached_property.py", line 124, in _get_ip_key_from_instance_cache
    instance_cache = self._get_instance_cache(instance)
  File "/code/mathesar/state/cached_property.py", line 138, in _get_instance_cache
    return instance.__dict__

Exception Type: AttributeError at /api/db/v0/tables/1/constraints/
Exception Value: 'NoneType' object has no attribute '__dict__'
```

**Review by dmos62 — APPROVED** [review]
Looks awesome! Well done. I only skimmed over the permission stuff (and glanced what seemed like valuable simplifications!). I'll leave it to Mukesh to do a thorough review of that.

**Comment by pavish** [comment]
@silentninja Here are the changes in existing contracts:

**Scenario**: User does not have access to table with <table_id>

**APIs**: GET requests to `tables/<table_id>/columns/` and `tables/<table_id>/constraints/`

**Previous behaviour**: APIs return empty list with status code 200.

**Behaviour in this PR**: APIs returns status code 403.

The failing test cases are modified to address this change, in this commit: https://github.com/centerofci/mathesar/pull/3092/commits/041b0c9cb87dd884b37f44645f2de8c4cb9efcbd

**Review by silentninja — APPROVED** [review]
Looks good to me!

File: mathesar/api/db/permissions/columns.py
Comment (pavish) [reply mathesar/api/db/permissions/columns.py:30]: I'd like to add a point to this.

The scope_queryset methods that were removed here need are not needed for table-level permissions, but for column and constraint level permission checks, if & when we have them.

Table level permission checks are handled at the table access policy which contains a scope_queryset method.

**Comment by pavish** [comment]
@dmos62 I'd like to merge this PR in as I have approval from you and Mukesh.

I'll add the tests for the new endpoints in an upcoming PR.

**Pull request merged**
